File: frontend/src/services/taskService.js
import api from './api';

/**
 * Task API service layer.
 * Abstracts all task-related API calls from components.
 */
const taskService = {
  /**
   * Fetches the 5 most recent non-completed tasks
   * @returns {Promise} List of tasks
   */
  getRecentTasks: async () => {
    try {
      const response = await api.get('/tasks');
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to fetch tasks');
    }
  },

  /**
   * Creates a new task
   * @param {Object} taskData - Task data with title and description
   * @returns {Promise} Created task
   */
  createTask: async (taskData) => {
    try {
      const response = await api.post('/tasks', taskData);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to create task');
    }
  },

  /**
   * Marks a task as completed
   * @param {number} taskId - Task ID
   * @returns {Promise} Updated task
   */
  completeTask: async (taskId) => {
    try {
      const response = await api.put(`/tasks/${taskId}/complete`);
      return response.data;
    } catch (error) {
      throw new Error(error.response?.data?.message || 'Failed to complete task');
    }
  },
};

export default taskService;

File: backend/src/main/java/com/todoapp/backend/service/TaskService.java
package com.todoapp.backend.service;

import com.todoapp.backend.dto.TaskCreateRequest;
import com.todoapp.backend.dto.TaskResponse;

import java.util.List;

/**
 * Service interface for Task business logic.
 * Follows Interface Segregation Principle (SOLID).
 */
public interface TaskService {
    
    /**
     * Creates a new task.
     * @param request Task creation request
     * @return Created task response
     */
    TaskResponse createTask(TaskCreateRequest request);
    
    /**
     * Retrieves the 5 most recent non-completed tasks.
     * @return List of up to 5 recent active tasks
     */
    List<TaskResponse> getRecentTasks();

    /**
     * Deletes all tasks - for testing purposes.
     */
    void deleteAllTasks();
    
    /**
     * Marks a task as completed.
     * @param id Task ID
     * @return Updated task response
     */
    TaskResponse completeTask(Long id);
}